Source/Task/WaitTimer_stl.cpp

```diff
@@ -7,88 +7,194 @@ class WaitTimerImpl
 {
 public:
 
-    ~WaitTimerImpl()
-    {
-        if (m_thread.joinable())
-        {
-            std::lock_guard<std::mutex> lock(m_mutex);
-            m_abortThread = true;
-            m_cv.notify_all();
-        }
+    HRESULT Initialize(_In_opt_ void* context, _In_ WaitTimerCallback* callback);
+    void Start(_In_ uint64_t absoluteTime);
+    void Cancel();
+    void InvokeCallback();
 
-        m_thread.join();
-    }
+private:
+
+    void* m_context;
+    WaitTimerCallback* m_callback;
+};
+
+struct TimerEntry
+{
+    Deadline When;
+    WaitTimerImpl* Timer;
+    TimerEntry(Deadline d, WaitTimerImpl* t) : When{ d }, Timer{ t } {}
+};
 
-    HRESULT Initialize(_In_opt_ void* context, _In_ WaitTimerCallback* callback)
+struct TimerEntryComparator
+{
+    bool operator()(TimerEntry const& l, TimerEntry const& r) noexcept
     {
-        m_context = context;
-        m_callback = callback;
+        return l.When > r.When;
+    }
+};
+
+class TimerQueue
+{
+public:
+    bool LazyInit() noexcept;
+
+    void Set(WaitTimerImpl* timer, Deadline deadline) noexcept;
+    void Remove(WaitTimerImpl const* timer) noexcept;
+
+private:
+    void Worker() noexcept;
+
+    TimerEntry const& Peek() const noexcept;
+    TimerEntry Pop() noexcept;
 
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<TimerEntry> m_queue; // used as a heap
+    bool m_initialized = false;
+};
+
+namespace
+{
+    std::once_flag g_timerQueueLazyInit;
+    TimerQueue g_timerQueue;
+}
+
+bool TimerQueue::LazyInit() noexcept
+{
+    std::call_once(g_timerQueueLazyInit, [this]()
+    {
         try
         {
-            std::thread t([this] { Worker(); });
-            m_thread.swap(t);
+            std::thread t([this]()
+            {
+                Worker();
+            });
+            t.detach();
+            m_initialized = true;
         }
         catch (...)
         {
-            return E_FAIL;
+            m_initialized = false;
         }
+    });
 
-        return S_OK;
-    }
+    return m_initialized;
+}
 
-    void Start(_In_ uint64_t absoluteTime)
+void TimerQueue::Set(WaitTimerImpl* timer, Deadline deadline) noexcept
+{
     {
-        std::lock_guard<std::mutex> lock(m_mutex);
-        m_duration = Deadline::duration(absoluteTime);
-        m_deadlineSet = true;
-        m_cv.notify_all();
-    }
+        std::lock_guard<std::mutex> lock{ m_mutex };
 
-    void Cancel()
-    {
-        std::lock_guard<std::mutex> lock(m_mutex);
-        m_deadlineSet = false;
-        m_cv.notify_all();
+        auto it = std::find_if(m_queue.begin(), m_queue.end(), [timer](auto &item) { return item.Timer == timer; });
+        if (it != m_queue.end())
```

Comment:
do you have unit tests checking this? It's been too long since I studied heaps to remember if this is a thing that is valid to do (or if it breaks invariants that push_heap relies on)

Reply:
You may want to do what we do in remove instead, and just clear the timer pointer in the old entry and then add a new entry